protocol/v1/status_message.go

```diff
@@ -160,165 +161,76 @@ func (m *StatusMessage) HandleApplicationMetadata() error {
-			log.Printf("[message::DecodeMessage] could not decode SyncInstallation: %#x, err: %v", m.Hash, err.Error())
-		} else {
-			m.ParsedMessage = message
+		return m.unmarshalProtobufData(new(protobuf.SyncInstallation))
 
-			return nil
-		}
 	case protobuf.ApplicationMetadataMessage_SYNC_INSTALLATION_CONTACT:
-		var message protobuf.SyncInstallationContact
 		log.Printf("Sync installation contact")
-		err := proto.Unmarshal(m.DecryptedPayload, &message)
-		if err != nil {
-			m.ParsedMessage = nil
-			log.Printf("[message::DecodeMessage] could not decode SyncInstallationContact: %#x, err: %v", m.Hash, err.Error())
-		} else {
-			m.ParsedMessage = message
+		return m.unmarshalProtobufData(new(protobuf.SyncInstallationContact))
 
-			return nil
-		}
 	case protobuf.ApplicationMetadataMessage_SYNC_INSTALLATION_PUBLIC_CHAT:
-		var message protobuf.SyncInstallationPublicChat
-		err := proto.Unmarshal(m.DecryptedPayload, &message)
-		if err != nil {
-			m.ParsedMessage = nil
-			log.Printf("[message::DecodeMessage] could not decode SyncInstallationPublicChat: %#x, err: %v", m.Hash, err.Error())
-		} else {
-			m.ParsedMessage = message
+		return m.unmarshalProtobufData(new(protobuf.SyncInstallationPublicChat))
 
-			return nil
-		}
 	case protobuf.ApplicationMetadataMessage_SYNC_INSTALLATION_ACCOUNT:
-		var message protobuf.SyncInstallationAccount
-		err := proto.Unmarshal(m.DecryptedPayload, &message)
-		if err != nil {
-			m.ParsedMessage = nil
-			log.Printf("[message::DecodeMessage] could not decode SyncInstallationAccount: %#x, err: %v", m.Hash, err.Error())
-		} else {
-			m.ParsedMessage = message
+		return m.unmarshalProtobufData(new(protobuf.SyncInstallationAccount))
 
-			return nil
-		}
 	case protobuf.ApplicationMetadataMessage_PAIR_INSTALLATION:
-		var message protobuf.PairInstallation
-		err := proto.Unmarshal(m.DecryptedPayload, &message)
-		if err != nil {
-			m.ParsedMessage = nil
-			log.Printf("[message::DecodeMessage] could not decode PairInstallation: %#x, err: %v", m.Hash, err.Error())
-		} else {
-			m.ParsedMessage = message
+		return m.unmarshalProtobufData(new(protobuf.PairInstallation))
 
-			return nil
-		}
+	case protobuf.ApplicationMetadataMessage_EMOJI_REACTION:
+		return m.unmarshalProtobufData(new(protobuf.EmojiReaction))
+
+	case protobuf.ApplicationMetadataMessage_EMOJI_REACTION_RETRACTION:
+		return m.unmarshalProtobufData(new(protobuf.EmojiReactionRetraction))
+
+	}
+	return nil
+}
+
+func (m *StatusMessage) unmarshalProtobufData(pb proto.Message) error {
+	var ptr proto.Message
+	rv := reflect.ValueOf(pb)
+	if rv.Kind() == reflect.Ptr{
+		ptr = pb
+	} else {
+		ptr = rv.Addr().Interface().(proto.Message)
 	}
+
+	err := proto.Unmarshal(m.DecryptedPayload, ptr)
+	if err != nil {
+		m.ParsedMessage = nil
+		log.Printf("[message::DecodeMessage] could not decode %T: %#x, err: %v", pb, m.Hash, err.Error())
+	} else {
+		rv = reflect.ValueOf(ptr)
+		m.ParsedMessage = rv.Elem()
+		return nil
+	}
+
 	return nil
 }
```

Comment:
I haven't yet implemented any tests on this func, and I'm not 100% is will always parse pointers and values as intended. It should work fine, but I'm not 100%